File: CS/MyListBoxDragDropManager.cs
// Developer Express Code Central Example:
// ListBoxDragDropManager - How to reorder items
// 
// The current ListBoxDragDropManager version does not provide the capability to
// reorder items. This example demonstrates how to implement this functionality
// manually.
// In this example, we have created a ListBoxDragDropManager class
// descendant and overridden its OnDragOver and OnDrop methods to add the
// capability to drop an item before or after another item.
// You can use this class
// like the original ListBoxDragDropManager in the following
// manner:
// 
// <dxe:ListBoxEdit x:Name="editor1"
// DisplayMember="Name">
// <i:Interaction.Behaviors>
// <local:MyListBoxDragDropManager
// x:Name="manager1"/>
// </i:Interaction.Behaviors>
// </dxe:ListBoxEdit>
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4598

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpf.Grid;
using DevExpress.Xpf.Editors;
using DevExpress.Xpf.Core.Native;
using System.Windows.Media;
using System.Reflection;
using DevExpress.Xpf.Grid.DragDrop;
using System.Windows;
using System.Windows.Input;
using System.Collections;
using System.ComponentModel;

namespace DXEditorsSample {
	public class MyListBoxDragDropManager: ListBoxDragDropManager {
        class DragDropHitTestResult {
            public ListBoxEditItem Element { get; private set; }
            public HitTestResultBehavior CallBack(HitTestResult result) {
                ListBoxEditItem item = LayoutHelper.FindParentObject<ListBoxEditItem>(result.VisualHit);
                if (item != null) {
                    Element = item;
                    return HitTestResultBehavior.Stop;
                }
                return HitTestResultBehavior.Continue;
            }
        }
        ListBoxEditItem TargetItem;
        public override void OnDragOver(DragDropManagerBase sourceManager, UIElement source, Point pt) {
            ListBoxDragOverEventArgs e = RaiseDragOverEvent(sourceManager, pt, DropTargetType.None);
            DropEventIsLocked = e.Handled ? !e.AllowDrop : !AllowDrop || sourceManager.DraggingRows == null || sourceManager.DraggingRows.Count < 1;
            if (!DropEventIsLocked) {
                ProcessTargetItem(sourceManager, pt);
                return;
            }
        }
        protected virtual void ProcessTargetItem(DragDropManagerBase sourceManager, Point pt) {
            TargetItem = GetVisibleHitTestElement(pt);
            if (TargetItem == null) {
                sourceManager.ViewInfo.DropTargetRow = null;
                sourceManager.SetDropTargetType(DropTargetType.DataArea);
                ShowDropMarker(ListBox, TableDragIndicatorPosition.None);
                return;
            }
            var position = Mouse.GetPosition(TargetItem);
            double height = TargetItem.ActualHeight;
            if (position.Y < height / 2) {
                sourceManager.SetDropTargetType(DropTargetType.InsertRowsBefore);
                ShowDropMarker(TargetItem, TableDragIndicatorPosition.Top);
            }
            else {
                sourceManager.SetDropTargetType(DropTargetType.InsertRowsAfter);
                ShowDropMarker(TargetItem, TableDragIndicatorPosition.Bottom);
            }
            sourceManager.ViewInfo.DropTargetRow = TargetItem.Content;
        }
        protected virtual ListBoxEditItem GetVisibleHitTestElement(Point pt) {
            DragDropHitTestResult result = new DragDropHitTestResult();
            VisualTreeHelper.HitTest(ListBox, null, new HitTestResultCallback(result.CallBack), new PointHitTestParameters(pt));
            return result.Element;
        }
        protected override void OnDrop(DragDropManagerBase sourceManager, UIElement source, Point pt) {
            if (DropEventIsLocked) return;
            ListBoxDropEventArgs e = RaiseDropEvent(sourceManager);
            if (!e.Handled) {
                if (sourceManager.DraggingRows.Count > 0 && AllowDrop) {
                    ProcessTargetItem(e.SourceManager, pt);
                    IList currentSource = ItemsSource;
                    IList draggedFromSource = e.SourceManager.GetSource(null);
                    foreach (object obj in e.DraggedRows) {
                        if (TargetItem != null && obj == TargetItem.Content)
                            continue;
                        draggedFromSource.Remove(obj);
                        switch (e.SourceManager.ViewInfo.DropTargetType) {
                            case DropTargetType.DataArea:
                                currentSource.Add(obj);
                                break;
                            case DropTargetType.InsertRowsAfter:
                                currentSource.Insert(currentSource.IndexOf(TargetItem.Content) + 1, obj);
                                break;
                            case DropTargetType.InsertRowsBefore:
                                currentSource.Insert(Math.Max(currentSource.IndexOf(TargetItem.Content), 0), obj);
                                break;
                        }
                    }
                    RaiseDroppedEvent(sourceManager, e.DraggedRows);
                }
            }
            HideDropMarker();
        }
        protected override IList ItemsSource {
            get {
                if (ListBox.ItemsSource is ICollectionView) {
                    return ((ICollectionView)ListBox.ItemsSource).SourceCollection as IList;
                }
                return ListBox.ItemsSource as IList;
            }
        }
	}
}

File: VB/MyListBoxDragDropManager.vb
' Developer Express Code Central Example:
' ListBoxDragDropManager - How to reorder items
' 
' The current ListBoxDragDropManager version does not provide the capability to
' reorder items. This example demonstrates how to implement this functionality
' manually.
' In this example, we have created a ListBoxDragDropManager class
' descendant and overridden its OnDragOver and OnDrop methods to add the
' capability to drop an item before or after another item.
' You can use this class
' like the original ListBoxDragDropManager in the following
' manner:
' 
' <dxe:ListBoxEdit x:Name="editor1"
' DisplayMember="Name">
' <i:Interaction.Behaviors>
' <local:MyListBoxDragDropManager
' x:Name="manager1"/>
' </i:Interaction.Behaviors>
' </dxe:ListBoxEdit>
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E4598
Imports System
Imports DevExpress.Xpf.Grid
Imports DevExpress.Xpf.Editors
Imports DevExpress.Xpf.Core.Native
Imports System.Windows.Media
Imports DevExpress.Xpf.Grid.DragDrop
Imports System.Windows
Imports System.Windows.Input
Imports System.Collections
Imports System.ComponentModel

Namespace DXEditorsSample

    Public Class MyListBoxDragDropManager
        Inherits ListBoxDragDropManager

        Private Class DragDropHitTestResult

            Private _Element As ListBoxEditItem

            Public Property Element As ListBoxEditItem
                Get
                    Return _Element
                End Get

                Private Set(ByVal value As ListBoxEditItem)
                    _Element = value
                End Set
            End Property

            Public Function CallBack(ByVal result As HitTestResult) As HitTestResultBehavior
                Dim item As ListBoxEditItem = LayoutHelper.FindParentObject(Of ListBoxEditItem)(result.VisualHit)
                If item IsNot Nothing Then
                    Element = item
                    Return HitTestResultBehavior.Stop
                End If

                Return HitTestResultBehavior.Continue
            End Function
        End Class

        Private TargetItem As ListBoxEditItem

        Public Overrides Sub OnDragOver(ByVal sourceManager As DragDropManagerBase, ByVal source As UIElement, ByVal pt As Point)
            dragOverSourceManager = sourceManager
            Dim e As ListBoxDragOverEventArgs = RaiseDragOverEvent(sourceManager, pt, DropTargetType.None)
            DropEventIsLocked = If(e.Handled, Not e.AllowDrop, Not AllowDrop OrElse sourceManager.DraggingRows Is Nothing OrElse sourceManager.DraggingRows.Count < 1)
            If Not DropEventIsLocked Then
                ProcessTargetItem(sourceManager, pt)
                Return
            End If
        End Sub

        Protected Overridable Sub ProcessTargetItem(ByVal sourceManager As DragDropManagerBase, ByVal pt As Point)
            TargetItem = GetVisibleHitTestElement(pt)
            If TargetItem Is Nothing Then
                sourceManager.ViewInfo.DropTargetRow = Nothing
                sourceManager.SetDropTargetType(DropTargetType.DataArea)
                ShowDropMarker(ListBox, TableDragIndicatorPosition.None)
                Return
            End If

            Dim position = Mouse.GetPosition(TargetItem)
            Dim height As Double = TargetItem.ActualHeight
            If position.Y < height / 2 Then
                sourceManager.SetDropTargetType(DropTargetType.InsertRowsBefore)
                ShowDropMarker(TargetItem, TableDragIndicatorPosition.Top)
            Else
                sourceManager.SetDropTargetType(DropTargetType.InsertRowsAfter)
                ShowDropMarker(TargetItem, TableDragIndicatorPosition.Bottom)
            End If

            sourceManager.ViewInfo.DropTargetRow = TargetItem.Content
        End Sub

        Protected Overridable Function GetVisibleHitTestElement(ByVal pt As Point) As ListBoxEditItem
            Dim result As DragDropHitTestResult = New DragDropHitTestResult()
            Call VisualTreeHelper.HitTest(ListBox, Nothing, New HitTestResultCallback(AddressOf result.CallBack), New PointHitTestParameters(pt))
            Return result.Element
        End Function

        Protected Overrides Sub OnDrop(ByVal sourceManager As DragDropManagerBase, ByVal source As UIElement, ByVal pt As Point)
            If DropEventIsLocked Then Return
            Dim e As ListBoxDropEventArgs = RaiseDropEvent(sourceManager)
            If Not e.Handled Then
                If sourceManager.DraggingRows.Count > 0 AndAlso AllowDrop Then
                    ProcessTargetItem(e.SourceManager, pt)
                    Dim currentSource As IList = ItemsSource
                    Dim draggedFromSource As IList = e.SourceManager.GetSource(Nothing)
                    For Each obj As Object In e.DraggedRows
                        If TargetItem IsNot Nothing AndAlso obj Is TargetItem.Content Then Continue For
                        draggedFromSource.Remove(obj)
                        Select Case e.SourceManager.ViewInfo.DropTargetType
                            Case DropTargetType.DataArea
                                currentSource.Add(obj)
                            Case DropTargetType.InsertRowsAfter
                                currentSource.Insert(currentSource.IndexOf(TargetItem.Content) + 1, obj)
                            Case DropTargetType.InsertRowsBefore
                                currentSource.Insert(Math.Max(currentSource.IndexOf(TargetItem.Content), 0), obj)
                        End Select
                    Next

                    RaiseDroppedEvent(sourceManager, e.DraggedRows)
                End If
            End If

            HideDropMarker()
        End Sub

        Protected Overrides ReadOnly Property ItemsSource As IList
            Get
                If TypeOf ListBox.ItemsSource Is ICollectionView Then
                    Return TryCast(CType(ListBox.ItemsSource, ICollectionView).SourceCollection, IList)
                End If

                Return TryCast(ListBox.ItemsSource, IList)
            End Get
        End Property
    End Class
End Namespace
